[ty] Support basic narrowing with aliased conditional expressions (#24302)
astral-sh/ruff: crates/ty_python_semantic/resources/mdtest/narrow/aliased_conditions.md

```py
def _(x: int | None):
    is_none = x is None
    if bool(is_none):
        reveal_type(x)  # revealed: None
    if is_none if True else False:
        reveal_type(x)  # revealed: None
    if is_none == True:
        # TODO: it would be nice to support this case, but even direct narrowing doesn't work here
        reveal_type(x)  # revealed: int | None
    if (is_none,)[0]:
        # TODO: same as above
        reveal_type(x)  # revealed: int | None
    if y := is_none:
        # TODO: same as above
        reveal_type(x)  # revealed: int | None
    if (lambda: is_none)():
        # TODO: same as above
        reveal_type(x)  # revealed: int | None
```

## Attribute access alias

```py
class A:
    x: int | None
    b: bool

    def negate_b(self):
        self.b = not self.b

def _(a: A):
    is_none = a.x is None
    if is_none:
        reveal_type(a.x)  # revealed: None
    else:
        reveal_type(a.x)  # revealed: int

def _(a: A):
    # Attribute targets are not treated as aliases.
    # It is difficult to track them accurately.
    a.b = a.x is None
    a.negate_b()
    if a.b:
        reveal_type(a.x)  # revealed: int | None
    else:

>         reveal_type(a.x)  # revealed: int | None

Reviewer: 👍

```

## Subscript access alias

```py
def _(l: list[int | None]):
    is_none = l[0] is None
    if is_none:
        reveal_type(l[0])  # revealed: None
    else:
        reveal_type(l[0])  # revealed: int

def _(l: list[int | None], lb: list[bool]):
    # Same as attributes: subscript targets are not treated as aliases.
    lb[0] = l[0] is None
    if lb[0]:
        reveal_type(l[0])  # revealed: int | None
    else:
        reveal_type(l[0])  # revealed: int | None
```

> ## Narrowed variable reassigned invalidates alias

Reviewer: Maybe

```suggestion
## Narrowing is invalidated when target is reassigned
```

> If the narrowed variable is reassigned between the alias and its use as a condition, the alias must
> not be followed.

Reviewer: ```suggestion
If the target is reassigned between the definition of the alias and its use as a condition,
narrowing does not take place:
```

    is_none = x is None
    if cond:
        x = 1
    if is_none:
        reveal_type(x)  # revealed: int | None

    is_none = x is None
    if is_none:
        reveal_type(x)  # revealed: None

class A:
    x: int | None

def _(a: A):
    is_none = a.x is None
    a.x = 1
    if is_none:
        reveal_type(a.x)  # revealed: Literal[1]

def _(a: A):
    is_none = a.x is None
    a = A()
    if is_none:
        reveal_type(a.x)  # revealed: int | None
```

## Alias variable reassigned invalidates alias

If the alias variable itself is reassigned, it no longer represents the original check.

```py
def _(x: int | None):
    is_none = x is None
    is_none = True
    if is_none:
        reveal_type(x)  # revealed: int | None

    is_none = x is None
    if is_none:
        reveal_type(x)  # revealed: None
```

## Nested scope can preserve alias

> TODO: This feature is not supported yet.

Aliases remain valid inside nested scopes. For eager scopes (class bodies, comprehensions) they are
evaluated inline. For lazy scopes (inner functions), the narrowing evaluator uses lazy snapshots to
track whether the narrowed variable was reassigned, so alias-based narrowing works the same as

> direct narrowing across scope boundaries.

Reviewer: Minor: I think this should focus less on implementation details, and rather describe the desired semantics for eager and lazy scopes.

```py
def _(x: int | None):
    is_none = x is None

    if is_none:
        reveal_type(x)  # revealed: None

>     class Inner:
>         if is_none:
>             # TODO: should be `None`
>             reveal_type(x)  # revealed: int | None
> 
>         def inner():
>             if is_none:
>                 # TODO: should be `None`
>                 reveal_type(x)  # revealed: int | None

Reviewer: Maybe this would help a bit with readability (similar below):

```suggestion
    class EagerScope:
        if is_none:
            # TODO: should be `None`
            reveal_type(x)  # revealed: int | None

        def lazy_scope():
            if is_none:
                # TODO: should be `None`
                reveal_type(x)  # revealed: int | None
```

    def inner2():
        if is_none:
            # TODO: should be `None`
            reveal_type(x)  # revealed: int | None

        class Inner2:
            if is_none:
                # TODO: should be `None`
                reveal_type(x)  # revealed: int | None

class A:
    x: int | None

def _(a: A):
    a = A()

>     a.x = None

Reviewer: This seems strange to me. Everything below should also work without this explicit assignment, right? With this assignment in place, it looks like narrowing by attribute assignment might interfere with what's actually being tested here.

In fact, if I remove this line, one more TODO needs to be added below (in the first case inside `Inner`), which makes sense to me (for symmetry reasons).

    is_none = a.x is None

    if is_none:
        reveal_type(a.x)  # revealed: None

    class Inner:
        if is_none:
            reveal_type(a.x)  # revealed: None

        def inner():
            if is_none:
                # TODO: should be `None`
                reveal_type(a.x)  # revealed: int | None

    def inner2():
        if is_none:
            # TODO: should be `None`
            reveal_type(a.x)  # revealed: int | None

        class Inner2:
            if is_none:
                # TODO: should be `None`
                reveal_type(a.x)  # revealed: int | None
```

## Cross-scope invalidation

> ### Narrowed variable reassigned

Reviewer: ```suggestion
### Target reassignments
```

> If the narrowed variable is reassigned inside an eager scope, the alias is invalidated within that
> scope.

Reviewer: Similar to above:

```suggestion
If the target is reassigned inside an eager scope, narrowing does not take place within that scope.
```

        class Inner2:
            # `is_none` here refers to the function scope variable, not the class-scope
            # `is_none = True`. Python's name resolution skips class scopes for nested
            # scopes, so the alias remains valid.
            if is_none:
                # TODO: should be `None`
                reveal_type(x)  # revealed: int | None

    if is_none:
        reveal_type(x)  # revealed: None
```

The same applies to a lazy scope:

```py
def _(x: int | None):
    is_none = x is None

    def inner():
        nonlocal is_none
        is_none = True
        if is_none:
            reveal_type(x)  # revealed: int | None

    inner()

    # TODO: should be `int | None`
    # We don't yet track that `inner()` can modify `is_none` via `nonlocal`.
    # (https://github.com/astral-sh/ty/issues/2731)
    if is_none:
        reveal_type(x)  # revealed: None

def _(x: int | None):
    is_none = x is None

    def inner():
        if is_none:
            reveal_type(x)  # revealed: int | None

        def inner2():
            if is_none:
                reveal_type(x)  # revealed: int | None

    is_none = True

    inner()
```

> ## Chained alias

Reviewer: Maybe
```suggestion
## Chained aliases
```
